DOC Mentioned efficiency and precision issues in glossary.rst

suhasid098 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
I found a TODO in glossary.rst that demanded to describe the efficiency and precision. It was documentation

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

suhasid098 committed:
Mentiond fficiency and prcision issues in glossary.rst

suhasid098 committed:
Merge branch 'main' into suhaBranch1

suhasid098 commented:
Hey @glemaitre saw that you mentioned we could ping you for PRs wondering if I could get a review please! 😃

glemaitre commented on doc/glossary.rst:
```diff
@@ -285,6 +285,16 @@ General Concepts
 
+        In regards to the efficiency and precision of the NumPy array, the data type
+        (`dtype`) can make a significant impact. If one chooses a data type with
+        a higher precision like `np.float64` or `np.int64` it will result in slow
```
I think that we need to mitigate the `slow operations`. Basically, using lower precision can provide a speed up if the operation under the hood leverage some optimizations as vectorization, SIMD, or cache optimization. If those are not implemented or taking care, you might not get any speed up.

glemaitre commented on doc/glossary.rst:
```diff
@@ -285,6 +285,16 @@ General Concepts
+        In regards to the efficiency and precision of the NumPy array, the data type
+        (`dtype`) can make a significant impact. If one chooses a data type with
+        a higher precision like `np.float64` or `np.int64` it will result in slow
+        operations, and increase memory usage, but it will result in accurate results.
```
accurate results -> more accurate results due to the ability to represent floating number with a lower floating point error.

glemaitre commented on doc/glossary.rst:
```diff
@@ -285,6 +285,16 @@ General Concepts
+        (`dtype`) can make a significant impact. If one chooses a data type with
+        a higher precision like `np.float64` or `np.int64` it will result in slow
+        operations, and increase memory usage, but it will result in accurate results.
+        However, if one opts for lower precision types like `np.float32` or `np.int32`
```
I don't thin that we need to repeat. We can just explicitly make the comparison between 64-bits and 32-bits numbers in the first sentence.

glemaitre commented on doc/glossary.rst:
```diff
@@ -285,6 +285,16 @@ General Concepts
+        operations, and increase memory usage, but it will result in accurate results.
+        However, if one opts for lower precision types like `np.float32` or `np.int32`
+        it would result in faster operations, and lower memory usage, but less accurate
+        results. The analysis of the choice will be dependent on the size of the 
+        dataset needed in machine learning tasks. For example, if a dataset is large,
+        it would be worth considering a lower-precision data type for a larger dataset
+        since speed and accuracy would be a priority.
```
Instead of speaking about datataset, it all comes to if the algorithm used under the hood is leveraging `np.float32`. For instance, some minimization method are only coded in `np.float64` and even passing `np.float32` will trigger a conversion and thus the code will be slower and more memory consumer in `np.float32` due to the extra conversion.

glemaitre replied:
You can remove the TODO below.

suhasid098 committed:
DOC Mentioned efficiency and precision issues in glossary.rst 

I've incorporated the valuable insights and suggestions from glemaitre to enhance the clarity and technical accuracy of the document. Here's a summary of the implemented changes:

Mitigating Slow Operations: As glemaitre suggested, we explored how using lower precision (e.g., np.float32 over np.float64) can speed up operations if the underlying process leverages optimizations like vectorization, SIMD, or cache optimization. This acknowledgment led to a more nuanced discussion on when and why these optimizations matter and how their absence could negate expected speed improvements.

Enhancing Accuracy Description: I updated the text to reflect the point about "more accurate results" being achievable with higher precision types due to their lower floating-point error. This change underscores the precision-efficiency trade-off more clearly.

Streamlined Comparison: Following the advice to avoid repetition, the initial comparison between 64-bit and 32-bit data types was made more explicit in the first sentence. This direct approach simplifies the narrative by immediately setting the stage for the discussion on trade-offs.

Algorithm Dependency: I shifted the focus from dataset size to the relevance of the algorithm's compatibility with np.float32. This section now highlights how some algorithms, particularly certain minimization methods, may not benefit from lower precision due to inherent coding in np.float64, leading to inadvertent conversions that can slow down the process and increase memory usage.

Removal of the TODO Section: As suggested, the TODO prompting further discussion on efficiency, precision, and casting policy has been removed to present the analysis as complete in its current form.

suhasid098 committed:
Merge branch 'main' into suhaBranch1

glemaitre committed:
fix make text less chatgpt-like phrasing

glemaitre reviewed (approved):
I removed the trailing spaces and rephrase a part that really look like written by ChatGPT with overly complicated (and not consistent with the rest of the document) choice of words.

glemaitre commented:
Enabling auto-merge. Thanks @suhasid098
